onnxruntime/featurizers_ops/cpu/forecasting_pivot_transformer.cc

```diff
@@ -0,0 +1,114 @@
+#include "core/framework/data_types.h"
+#include "core/framework/data_types_internal.h"
+#include "core/framework/op_kernel.h"
+
+#include "Featurizers/ForecastingPivotFeaturizer.h"
+#include "Featurizers/../Archive.h"
+
+namespace NS = Microsoft::Featurizer;
+namespace onnxruntime {
+namespace featurizers {
+
+template <typename T> //float, double
+struct ForecastingPivotTransformerImpl {
+  void operator()(OpKernelContext* ctx) const {
+    using TransformerT = Microsoft::Featurizer::Featurizers::ForecastingPivotTransformer<T>;
+    using MatrixT = NS::RowMajMatrix<T>;
+    using InputMatrixT = Eigen::Map<const MatrixT>;
+    using InputType = std::vector<InputMatrixT>;
+    using OutputType = std::vector<T>;
+
+    //Get the transformer
+    const auto* state_tensor(ctx->Input<Tensor>(0));
+    const uint8_t* const state_data(state_tensor->Data<uint8_t>());
+    Microsoft::Featurizer::Archive archive(state_data, state_tensor->Shape().GetDims()[0]);
+    TransformerT transformer(archive);
+
+    // Get the Number of Rows
+    const auto* input_tensor_temp(ctx->Input<Tensor>(1));
+    const int64_t row_num = input_tensor_temp->Shape()[0];
+
+    //Get the output for whole rows is inevitable because there is conceptually no way to determine the shape of output for each row
+    std::vector<OutputType> output;
+    std::function<void(OutputType const & value)> callback_fn;
+    callback_fn = [&output](OutputType const & value) -> void {
+      output.emplace_back(value);
+    };
+
+    // Transform
+    const int input_node_0_count = ctx->NumVariadicInputs(0);
+    const int input_node_1_count = ctx->NumVariadicInputs(1);
+    InputType input;
+    input.reserve(input_node_1_count);
+    std::unordered_map<int, std::tuple<const T*,int64_t, int64_t>> dataPtrMap;
+    for (int64_t row_idx = 0; row_idx < row_num; ++row_idx) {
+      //Prepare Input and Output
+      input.clear();
+      for (int index = input_node_0_count; index < input_node_0_count + input_node_1_count; ++index) {
+        if (row_idx == 0) {
+          //Get the Input
+          const auto* input_tensor(ctx->Input<Tensor>(index));
+          const T* input_data(input_tensor->template Data<T>());
+          // Matrix Eigen raw buffer mapping
+          const int64_t input_dim_1 = input_tensor->Shape()[1];
+          const int64_t input_dim_2 = input_tensor->Shape()[2];
+          //store data pointer and dimension information
+          std::tuple<const T*,int64_t, int64_t> info_tuple(input_data, input_dim_1, input_dim_2);
+          dataPtrMap.insert(std::pair<int, std::tuple<const T*,int64_t, int64_t>>(index, info_tuple));
+        }
+        const T* input_data(std::get<0>(dataPtrMap.at(index)));
+        const int64_t input_dim_1(std::get<1>(dataPtrMap.at(index)));
+        const int64_t input_dim_2(std::get<2>(dataPtrMap.at(index)));
+        InputMatrixT input_matrix(input_data, input_dim_1, input_dim_2);
+        input.emplace_back(input_matrix);
```

Comment:
>emplace_back [](start = 14, length = 12)

emplace was created so you donot need to create the value there and copy it. It can take ctor arguments to create input_matrix inplace.

Reply:
.emplace(input_data, input_dim_1, input_dim_2)

---
In reply to: [405122418](https://github.com/microsoft/onnxruntime/pull/3435#discussion_r405122418) [](ancestors = 405122418)